### PR TITLE
Fix entity leak in SaveLoadSaveTest, StackTest cleanup

### LINKED ISSUE #45078: MergeTest test failure on main (Content.IntegrationTests.Tests.Stacks.StackTest.MergeTest)
## Test Failure On Main

A test failure has been detected on the main branch. This should never occur.

> [!NOTE]
> This issue is automatically generated by a github action and will be updated when this failure reoccurs.
> Do not edit the body of this issue.

## Failure Details
<dl>
    <dt>Test name</dt>
    <dd>MergeTest</dd>
    <dt>Fully qualified name</dt>
    <dd>Content.IntegrationTests.Tests.Stacks.StackTest.MergeTest</dd>
    <dt>Last encountered</dt>
    <dd>Friday, August 7th</dd>
</dl>

<details>
<summary>Failure message</summary>

```
Multiple failures or warnings in test:
  1)   Assert.That(SEntMan.EntityCount, Is.EqualTo(1))
  Expected: 1
  But was:  8

     at Content.IntegrationTests.Tests.Stacks.StackTest.MergeTest() in /home/runner/work/space-station-14/space-station-14/Content.IntegrationTests/Tests/Stacks/StackTests.cs:line 69
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)

  2) Test was dirty-disposed.
     at Robust.UnitTesting.Pool.TestPair`2.OnDirtyDispose() in /home/runner/work/space-station-14/space-station-14/RobustToolbox/Robust.UnitTesting/Pool/TestPair.Recycle.cs:line 44
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Robust.UnitTesting.Pool.TestPair`2.OnDirtyDispose()
   at Robust.UnitTesting.Pool.TestPair`2.DisposeAsync() in /home/runner/work/space-station-14/space-station-14/RobustToolbox/Robust.UnitTesting/Pool/TestPair.Recycle.cs:line 146
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Robust.UnitTesting.Pool.TestPair`2.DisposeAsync()
   at Content.IntegrationTests.Fixtures.GameTest.DoTeardown() in /home/runner/work/space-station-14/space-station-14/Content.IntegrationTests/Fixtures/GameTest.cs:line 263
   at Content.IntegrationTests.Fixtures.GameTest.DoTeardown()
   at InvokeStub_GameTest.DoTeardown(Object, Object, IntPtr*)
```

</details>

<details>
<summary>Test output</summary>

```
GetPair: Called by test Stacks.StackTest.MergeTest
GetPair: Looking in pool for a suitable pair
GetPair: Suitable pair found (Pair 3)
GetPair: Cleanup not needed, Skipping cleanup of pair
SERVER: 0.000s [WARN] cfg: Only clients can change the 'net.interp' cvar.
SERVER: 0.000s [WARN] cfg: Robust.UnitTesting.RobustIntegrationTest+IntegrationNetManager+IntegrationNetChannel: Received late nwVar message (1164 < 1168 ).
GetPair: Retrieving pair 3 from pool took 2.4028 ms
GetPair: Returning pair 3 with client/server seeds: 321269909/460670777
DisposeAsync: Dirty return of pair 3 started
DisposeAsync: Test gave back pair 3 in 16.0957 ms
DisposeAsync: Disposed pair 3 in 0.9851 ms
```

</details>

## Using this issue
Please place troubleshooting notes in the comments of this issue.

### LINKED ISSUE #45081: MergeOverflowTest test failure on main (Content.IntegrationTests.Tests.Stacks.StackTest.MergeOverflowTest)
## Test Failure On Main

A test failure has been detected on the main branch. This should never occur.

> [!NOTE]
> This issue is automatically generated by a github action and will be updated when this failure reoccurs.
> Do not edit the body of this issue.

## Failure Details
<dl>
    <dt>Test name</dt>
    <dd>MergeOverflowTest</dd>
    <dt>Fully qualified name</dt>
    <dd>Content.IntegrationTests.Tests.Stacks.StackTest.MergeOverflowTest</dd>
    <dt>Last encountered</dt>
    <dd>Sunday, August 16th</dd>
</dl>

<details>
<summary>Failure message</summary>

```
Multiple failures or warnings in test:
  1)   Assert.That(SEntMan.EntityCount, Is.EqualTo(2))
  Expected: 2
  But was:  9

     at Content.IntegrationTests.Tests.Stacks.StackTest.MergeOverflowTest() in /home/runner/work/space-station-14/space-station-14/Content.IntegrationTests/Tests/Stacks/StackTests.cs:line 108
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)

  2) Test was dirty-disposed.
     at Robust.UnitTesting.Pool.TestPair`2.OnDirtyDispose() in /home/runner/work/space-station-14/space-station-14/RobustToolbox/Robust.UnitTesting/Pool/TestPair.Recycle.cs:line 44
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Robust.UnitTesting.Pool.TestPair`2.OnDirtyDispose()
   at Robust.UnitTesting.Pool.TestPair`2.DisposeAsync() in /home/runner/work/space-station-14/space-station-14/RobustToolbox/Robust.UnitTesting/Pool/TestPair.Recycle.cs:line 146
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Robust.UnitTesting.Pool.TestPair`2.DisposeAsync()
   at Content.IntegrationTests.Fixtures.GameTest.DoTeardown() in /home/runner/work/space-station-14/space-station-14/Content.IntegrationTests/Fixtures/GameTest.cs:line 263
   at Content.IntegrationTests.Fixtures.GameTest.DoTeardown()
   at InvokeStub_GameTest.DoTeardown(Object, Object, IntPtr*)
```

</details>

<details>
<summary>Test output</summary>

```
GetPair: Called by test Stacks.StackTest.MergeOverflowTest
GetPair: Looking in pool for a suitable pair
GetPair: Suitable pair found (Pair 3)
GetPair: Cleanup not needed, Skipping cleanup of pair
SERVER: 0.000s [WARN] cfg: Only clients can change the 'net.interp' cvar.
GetPair: Retrieving pair 3 from pool took 2.9184 ms
GetPair: Returning pair 3 with client/server seeds: 174075337/593968908
DisposeAsync: Dirty return of pair 3 started
DisposeAsync: Test gave back pair 3 in 19.5216 ms
DisposeAsync: Disposed pair 3 in 5.2149 ms
```

</details>

## Using this issue
Please place troubleshooting notes in the comments of this issue.

### PR DESCRIPTION
**Pull request by whatston3**
## About the PR

Trying to fix a failing test that I _think_ is leaking entities.  CreateSaveLoadSaveGrid was missing a map deletion at the end of its tests.

## Why / Balance

Tracking down the culprits for https://github.com/space-wizards/space-station-14/issues/45078.

Should resolve #45078.  (This time for sure!)
Should resolve #45081.

## Technical details

Shuffling of async events into tighter groups in StackTest (please scrutinize this).

Adding Assume checks at the top of the stack tests that care about entity count (they will fail if it's not zero, so it's not really a failure of the test, it's the environment that's wrong).

Adding another set of assumptions/assertions at the top and bottom of the map spawning tests, considering that they won't get cleaned up using the standard test cleanup functions.

## Test plan

Run tests!

Runs on my machine award, but hopefully if it fails, we can clean it up in this PR.

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have tested this pull request and written instructions on how to test it
- [X] I have added media to this PR or it does not require an in-game showcase.

## Changelog

### CODE REVIEW

File: Content.IntegrationTests/Tests/SaveLoadSaveTest.cs [hunk 1]
```diff
@@ -28,26 +28,30 @@ public async Task CreateSaveLoadSaveGrid()
...
                 var grid0 = mapSystem.CreateGridEntity(mapId0);
                 entManager.RunMapInit(grid0.Owner, entManager.GetComponent<MetaDataComponent>(grid0));
                 Assert.That(mapLoader.TrySaveGrid(grid0.Owner, rp1));
-                mapSystem.CreateMap(out var mapId1);
+                mapSystem.CreateMap(out mapId1);
                 Assert.That(mapLoader.TryLoadGrid(mapId1, rp1, out var grid1));
                 Assert.That(mapLoader.TrySaveGrid(grid1!.Value, rp2));
             });
 
-            await server.WaitIdleAsync();
```
Comment (kontakt) [inline Content.IntegrationTests/Tests/SaveLoadSaveTest.cs:50]: Why is this safe to remove?

Comment (whatston3) [reply Content.IntegrationTests/Tests/SaveLoadSaveTest.cs:50]: It's inherent in the waitPost.

[It's the wait part!]

**Review by kontakt — APPROVED** [review]

**Comment by kontakt** [comment]
This resolves the failures but it does cause the tests to be skipped as they fail assumptions. 
There is further cleanup to be done, but this is a good step imo.

**Comment by whatston3** [comment]
> This resolves the failures but it does cause the tests to be skipped as they fail assumptions. There is further cleanup to be done, but this is a good step imo.

I thought the assumptions still caused the tests to fail.  That's probably not ideal.

Also, I think the three stack tests that don't create a map _should_ be possible to be made server-sided tests, but I don't know if RunTicks alone is sufficient there.

[Maybe the build-test-debug.yml GHA should have `--failoninconclusive`?]

**Comment by kontakt** [comment]
> I thought the assumptions still caused the tests to fail. That's probably not ideal.
> 
> Also, I think the three stack tests that don't create a map _should_ be possible to be made server-sided tests, but I don't know if RunTicks alone is sufficient there.

Assumptions cause the test to be marked inconclusive as the test could not run properly. 
The failure should be reported by another test that's actually testing that functionality. 
We could make the GHA report those as failures, but ideally on things that actually fail are reported as failures.

For example, these tests shouldn't fail because a *different test* didn't clean up. It helps us understand what's actually causing problems.

**Review by Fildrance — APPROVED** [review]
This is a bit hectic - not full modernization. I am not toally sure why most of this cannot be also just done with [RunOnSide(Side.Server)] and not waiting for Server.WaitPost, but i assume there can be some strings attached ;( and PR mostly about fixing up map thingy so okay.
Cleaning up maps seems as right thing to do, in general.

**Comment by whatston3** [comment]
> This is a bit hectic - not full modernization. I am not toally sure why most of this cannot be also just done with [RunOnSide(Side.Server)] and not waiting for Server.WaitPost, but i assume there can be some strings attached ;( and PR mostly about fixing up map thingy so okay. Cleaning up maps seems as right thing to do, in general.

Can't remember at the moment.  Happy to have a followup that meets your standards if you'd prefer.  I'm not very familiar with writing these tests.

**Comment by Fildrance** [comment]
> Can't remember at the moment. Happy to have a followup that meets your standards if you'd prefer. I'm not very familiar with writing these tests.

Its okay, Sloth for example does not really want this new stuff, as its kinda redundant (GameTest) - its like 8-th or 11-th test template thing. And it is harder to debug then some other alternatives. So i am not gonna insist on it unless we will have hard commitment. We kinda had it at some point, but... oh well, not now i guess.
